File: old/lib/unionfind.cc
#include <vector>

struct UnionFind {

  typedef int Index;
  typedef Index Size;

  UnionFind(Index size): parent_of(size), size_of(size, 1), ngroups(size) {
    for (int i = 0; i < size; i++) {
      parent_of[i] = i;
    }
  }

  Index find(Index i) {
    if (parent_of[i] != i) {
      parent_of[i] = find(parent_of[i]);
    }
    return parent_of[i];
  }

  void join(Index raw_a, Index raw_b) {
    const Index a = find(raw_a);
    const Index b = find(raw_b);
    if (a == b) {
      return;
    }
    ngroups--;
    const Size ra = size_of[a];
    const Size rb = size_of[b];
    if (ra < rb) {
      parent_of[a] = b;
      size_of[b] += size_of[a];
    } else {
      parent_of[b] = a;
      size_of[a] += size_of[b];
    }
  }

  Index make_set() {
    Index i = parent_of.size();
    parent_of.push_back(i);
    size_of.push_back(1);
    return i;
  }

  Size size(Index i) {
    return size_of[find(i)];
  }

  Size number_of_nodes() const {
    return parent_of.size();
  }

  Size number_of_groups() const {
    return ngroups;
  }

private:
  std::vector<Index> parent_of;
  std::vector<Size> size_of;
  Size ngroups;
};

using namespace std;

#include <iostream>
#include <assert.h>

int main() {
  {
    UnionFind uf(10);
    assert(uf.find(1) != uf.find(2));
    assert(uf.find(1) == uf.find(1));
    assert(uf.size(1) == 1);
    assert(uf.number_of_nodes() == 10);
    assert(uf.number_of_groups() == 10);
    uf.join(1, 2);
    assert(uf.find(1) == uf.find(2));
    assert(uf.size(1) == 2);
    assert(uf.number_of_nodes() == 10);
    assert(uf.number_of_groups() == 9);
    uf.join(1, 2);
    assert(uf.find(1) == uf.find(2));
    assert(uf.size(1) == 2);
    assert(uf.number_of_nodes() == 10);
    assert(uf.number_of_groups() == 9);
  }
  cout << "pass" << endl;
}

File: old/lib/unionfind.py
"""unionfind.py"""

class UnionFind(object):

  def __init__(self, N):
    self._parent = list(range(N))
    self._size = [1] * N

  def find(self, i):
    if self._parent[i] != i:
      self._parent[i] = self.find(self._parent[i])
    return self._parent[i]

  def join(self, a, b):
    a = self.find(a)
    b = self.find(b)
    if a == b:
      return b
    if self._size[b] < self._size[a]:
      a, b = b, a

    self._size[b] += self._size[a]
    self._parent[a] = b
    return b

  def connected(self, a, b):
    return self.find(a) == self.find(b)

  def sizeof(self, i):
    return self._size[self.find(i)]

if __name__ == '__main__':
  uf = UnionFind(10)

  for i in range(10):
    assert uf.sizeof(i) == 1, uf.sizeof(i)
    for j in range(i+1, 10):
      assert not uf.connected(i, j)

  for i in range(1, 10):
    uf.join(1, i)

  assert uf.sizeof(0) == 1, uf.sizeof(0)
  for i in range(1, 10):
    assert uf.sizeof(i) == 9, (i, uf.sizeof(i))
    assert uf.find(i) == uf.find(1), uf.find(i)
    assert uf.connected(1, i)

  print('pass')
